File: backend/src/test/kotlin/org/globe42/web/activities/ActivityTypeControllerMvcTest.java
package org.globe42.web.activities;

import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.Arrays;

import org.globe42.dao.PersonDao;
import org.globe42.domain.ActivityType;
import org.globe42.domain.Person;
import org.globe42.test.GlobeMvcTest;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.test.web.servlet.MockMvc;

/**
 * MVC tests for {@link ActivityTypeController}
 * @author JB Nizet
 */
@GlobeMvcTest(ActivityTypeController.class)
public class ActivityTypeControllerMvcTest {
    @MockBean
    private PersonDao mockPersonDao;

    @Autowired
    private MockMvc mvc;

    @Test
    public void shouldListParticipants() throws Exception {
        Person person = new Person(42L);
        person.setEmail("john@doe.com");
        when(mockPersonDao.findParticipants(ActivityType.MEAL)).thenReturn(Arrays.asList(person));

        mvc.perform(get("/api/activity-types/{activityType}/participants", ActivityType.MEAL))
           .andExpect(status().isOk())
           .andExpect(jsonPath("$[0].id").value(42))
           .andExpect(jsonPath("$[0].email").value(person.getEmail()));
    }
}

File: backend/src/test/kotlin/org/globe42/web/activities/ActivityTypeControllerMvcTest.kt
package org.globe42.web.activities

import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import org.globe42.dao.PersonDao
import org.globe42.domain.ActivityType
import org.globe42.domain.Gender
import org.globe42.domain.Person
import org.globe42.test.GlobeMvcTest
import org.globe42.web.test.jsonValue
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get

/**
 * MVC tests for [ActivityTypeController]
 * @author JB Nizet
 */
@GlobeMvcTest(ActivityTypeController::class)
class ActivityTypeControllerMvcTest(@Autowired private val mvc: MockMvc) {
    @MockkBean
    private lateinit var mockPersonDao: PersonDao

    @Test
    fun `should list participants`() {
        val person = Person(42L, "John", "Doe", Gender.MALE)
        person.email = "john@doe.com"
        every { mockPersonDao.findParticipants(ActivityType.MEAL) } returns listOf(person)

        mvc.get("/api/activity-types/{activityType}/participants", ActivityType.MEAL).andExpect {
            status { isOk }
            jsonValue("$[0].id", 42)
            jsonValue("$[0].email", person.email!!)
        }
    }
}
